### ISSUE TITLE
RESEARCH/ENGINE: one-adjoint exact SAE outer gradient — Γ = ∇_θ log|H| collapses the third-order LAML correction to dot products

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1006]
## The theorem (completes #1004; supersedes its missing-term caveat)

Criterion: `V(ρ) = L(θ̂(ρ),ρ) + ½log|H(θ̂(ρ),ρ)| − occam(ρ)`, inner stationarity `g(θ̂,ρ)=0`, θ = (per-row gate/coordinate blocks, border β), H = the joint arrow Hessian **as the criterion computes it**.

Exact total derivative, all coordinates at once:

```
dV/dρ_j = ∂L/∂ρ_j − ∂occam/∂ρ_j + ½·tr(H⁻¹·∂H/∂ρ_j) − ½·Γᵀ·H⁻¹·(∂g/∂ρ_j)
Γ_a    := tr(H⁻¹·∂H/∂θ_a)      (the log-det adjoint w.r.t. INNER variables)
```

Derivation: envelope kills ∂L/∂θ·θ̂'; d log|H|/dρ_j = tr(H⁻¹[∂_ρj H + Σ_a ∂_θa H·θ̂'_a]) and θ̂'_j = −H⁻¹·∂g/∂ρ_j by IFT. The third-order tensor never appears as a tensor — only as the single adjoint vector Γ.

## Why this is cheap on the arrow structure

- **Γ is one row-pass.** ∂H/∂θ_a for a row-i variable touches only H_tt^(i), H_tβ^(i), and row i's border contribution. So Γ_a contracts the cached selected-inverse blocks — `(H⁻¹)_tt^(ii) = H_tt^(i)⁻¹ + H_tt^(i)⁻¹H_tβ^(i)·Schur⁻¹·H_tβ^(i)ᵀH_tt^(i)⁻¹`, `(H⁻¹)_tβ^(i) = −H_tt^(i)⁻¹H_tβ^(i)·Schur⁻¹`, and Schur⁻¹ — against per-row third-derivative channels of the row program (gate nonlinearity × basis jets × data fit). The per-row factor cache and Schur⁻¹ already exist at the evidence step (the EFS lane reads the same traces); the third-derivative channels are the #932 jet-tower / #683 axis-projected-tensor pattern applied to the SAE row program.
- **∂g/∂ρ_j is arrow-sparse per class**: ARD log-α_a → α·t̂ on one axis per row; log-λ_smooth → λ(S̃⊗I)β̂ on the border (plus S̃(θ) pullback cross-blocks, see landmine 2); log-λ_sparse → per-row gate-penalty gradient. Each u_j = H⁻¹∂g/∂ρ_j is one back-solve on the cached factor.
- **Explicit traces**: tr(H⁻¹∂H/∂λ_smooth) is the existing effective-dof trace; tr(H⁻¹∂H/∂λ_sparse) needs the per-row (H⁻¹)_tt diag blocks — same pass as Γ. This also fills the 'no sparse logdet trace' gap recorded by the #1004 FD-gate skeleton (3839bccb5).

Total cost: ~one extra criterion-evaluation-equivalent for the WHOLE gradient. Outer becomes BFGS (later exact-Newton via a second adjoint pass) with ~10-20 evaluations replacing ~200 compass polls + 12 seed evals; with capability (Analytic, Unavailable) the small-ρ SAE plan routes to gradient-BFGS and CompassSearch leaves the SAE path permanently (#1004's goal).

## Correctness landmines (contract, not caveats)

1. **Differentiate the criterion's H, not the true Hessian.** Where assembly uses a PSD majorizer for a block (SCAD curvature, entropy Gershgorin), log|H| is defined on the majorizer — Γ must differentiate THE SAME object on the same smooth branch, or value/gradient desync (the recurring bug class) returns. Single source of truth: derive each block's θ-derivative channel next to the block's assembly code (#931 criterion-atoms discipline).
2. **The #673 pullback-reparameterized roughness S̃(θ)**: its θ-dependence already contributes to the inner g; the SAME terms appear in ∂g/∂λ_smooth structure and in ∂H/∂θ. Verify against the inner assembly rather than assuming S̃ constant.
3. Gate-tier penalties at active-set boundaries (JumpReLU thresholds, IBP MAP kinks): Γ is defined on the smooth branch the criterion evaluates; document where one-sided.

## Verification ladder

1. FD-vs-analytic on the #1004 FD gate fixture (tests/sae_outer_gradient_fd_gate.rs) — the correction term must close the recorded residual exactly.
2. Same gate at a frames-active fixture (factored border).
3. Wall-time gate: gate-0-class K=2 fixture outer converges in ≤20 evaluations.

Builds on: #1004 (explicit terms + FD gate, landed 3839bccb5), #932 jet towers, #740 ψ-channel precedent, EFS selected-inverse traces. Closes the road to deleting CompassSearch from the SAE path.

**Comment by SauersML** [comment]
## Implementation spec: the per-row third-derivative channels for Γ (derivation)

Row model (Gaussian row, dispersion absorbed): `x̂_i(v, B) = Σ_k ζ(ℓ_k)·m_k(t_k)`, with `m_k(t) = B_kᵀΦ_k(t) ∈ ℝ^p`, gates `ζ(ℓ) = σ(ℓ/τ)·π_k` (ibp_map; softmax/jumprelu supply their own ζ, ζ', ζ'', ζ'''), row variables `v = (ℓ_{1..K}, t_{k,a})`, residual `r = x_i − x̂_i`.

**Model jets (all analytic, per basis kind):**
- `∂x̂/∂ℓ_k = ζ'_k·m_k`, `∂x̂/∂t_{k,a} = ζ_k·B_kᵀ(∂_aΦ_k)`
- `∂²x̂`: `ζ''_k·m_k` (ℓℓ), `ζ'_k·B_kᵀ∂_aΦ_k` (ℓt, same atom), `ζ_k·B_kᵀ∂²_{ab}Φ_k` (tt); cross-atom ≡ 0
- `∂³x̂`: `ζ'''_k·m_k`, `ζ''_k·B_kᵀ∂_aΦ`, `ζ'_k·B_kᵀ∂²Φ`, `ζ_k·B_kᵀ∂³_{abc}Φ_k` — third basis jets exist for harmonics analytically and for Duchon via the radial-power derivatives (the same family as the isometry third jets).

**Row Hessian (the FULL Hessian the evidence factor uses — it can be indefinite, which is consistent with the strict non-PD policy):**
`H_ab = (∂_a x̂)ᵀ(∂_b x̂) − Σ_c r_c·∂²_{ab}x̂_c + prior_ab` (priors: ARD diag α on t — quadratic, no third channel; gate-sparsity s(·) contributes its own s'' here and s''' below, on the same smooth branch / majorizer the criterion uses — landmine 1).

**The third channel (what Γ contracts):**
```
∂H_ab/∂w = (∂²_{aw}x̂)ᵀ(∂_b x̂) + (∂_a x̂)ᵀ(∂²_{bw}x̂)
         − Σ_c [ r_c·∂³_{abw}x̂_c − (∂_w x̂_c)·(∂²_{ab}x̂_c) ]
         + prior'''_{abw}
```
for `w` a row variable. For `w` a border coordinate (an entry of B_k, or of C_k through the frame lift): `x̂` is LINEAR in B, so replace `∂_w x̂` by the corresponding basis-gate product and drop one B from each term — `H` is quadratic in B, so these channels are exact products of existing first/second jets.

**Γ assembly (one row-pass):** with the cached undamped factor providing `Schur⁻¹` and per-row `H_tt^{(i)-1}`:
- `(H⁻¹)_{vv}^{(ii)} = H_vv^{(i)-1} + H_vv^{(i)-1}H_vβ^{(i)}·Schur⁻¹·H_vβ^{(i)ᵀ}H_vv^{(i)-1}`
- `(H⁻¹)_{vβ}^{(i)} = −H_vv^{(i)-1}H_vβ^{(i)}·Schur⁻¹`
- `Γ_w (row w) = tr[(H⁻¹)_{vv}^{(ii)}·∂H_vv^{(i)}/∂w] + 2·tr[(H⁻¹)_{vβ}^{(i)ᵀ}·∂H_vβ^{(i)}/∂w] + tr[Schur⁻¹·∂(row i's border Gram contribution)/∂w]`
- `Γ_w (border w)`: same shape, summed over rows that touch atom(w) — accumulate during the same pass.

Everything is `O(q_i³ + q_i²·r_loc)` per row plus small border contractions; at frames-active LLM scale the border pieces ride the factored layout from 13f556699.

**Then the gradient** (per #1006 header): `dV/dρ_j = explicit_j − ½·Γᵀ·H⁻¹·(∂g/∂ρ_j)`, with `∂g/∂ρ` per class: ARD `α·t̂` (one axis/row), λ_smooth `λ(S̃⊗I)β̂` (+ S̃(θ) cross-blocks per landmine 2), λ_sparse per-row `λ_sp·∇s`. Back-solves on the cached factor.

**Staging:** (1) Γ for the row-tier channels + λ_sparse explicit trace → close the FD-gate residual on the small fixture (tests/sae_outer_gradient_fd_gate.rs); (2) border channels + frames; (3) wire capability `(Analytic, Unavailable)` and retire CompassSearch from the SAE path. Stage 1's FD gate must close to machine-FD tolerance before stage 3 is allowed — desync discipline.

**Comment by SauersML** [comment]
Closed by the IBP empirical-π third channel landing on main.

**Commits**
- `a609a26e0` — wire the IBP-MAP `hessian_diag` logit third channel into `logdet_theta_adjoint`; IBP outer-ρ gradient is now `(Analytic)`.
- (penalty primitive) `IBPAssignmentPenalty::hessian_diag_logit_third_channels` — the exact `d/dℓ` of `hessian_diag`, including the new `sdd_k = ∂²score_k/∂M_k²` primitive.
- `08e657679` — direct Γ-vs-FD-of-`log|H|` gate for the IBP cross-row channel.
- `f3788712d` — H-consistency audit proof comment (see #1004 close-out / landmine 1 below).

**What was the gap.** `logdet_theta_adjoint`'s IBP-MAP branch returned a loudly-commented `0.0`, and `analytic_outer_rho_gradient_supported()` excluded IBP-MAP, because the assembled `htt` diagonal consumes `IBPAssignmentPenalty::hessian_diag`, whose logit derivative includes the plug-in stick-breaking MAP `π_k(M_k)` with `M_k = Σ_i z_ik` — a **column-global** coupling, not a row-local primitive.

**The derivation (on the criterion's own branch — landmine 1).** The assembled diagonal is `H_ik = w·[sd_k·J_ik² + score_k·c_ik]` with `J=z(1−z)/τ`, `c=z(1−z)(1−2z)/τ²`, `score_k`, `sd_k=∂score_k/∂M_k` exactly as `hessian_diag` assembles them. Its logit derivative splits cleanly:

```
∂H_ik/∂ℓ_wk = δ_iw·(∂_z H_ik)·J_ik   +   (∂_M H_ik)·J_wk
∂_z H_ik = w·J_ik·[ sd_k·2J_ik·(1−2z)/τ + score_k·(1−6z+6z²)/τ² ]   (row-local)
∂_M H_ik = w·[ sdd_k·J_ik² + sd_k·c_ik ]                            (cross-row, through π_k)
sdd_k = ∂sd_k/∂M_k = ∂²score_k/∂M_k².
```

`Γ_wk = Σ_i (H⁻¹)_ik,ik · ∂H_ik/∂ℓ_wk`. The row-local channel rides the existing `a==b==w` primitive (`assignment_prior_hdiag_derivative_entry`, deliberately one-sided to avoid double-counting); the cross-row channel is `C_k = Σ_i (H⁻¹)_ik,ik · ∂_M H_ik` distributed as `C_k · J_wk` to every active row `w` in a single column pass after the row loop. All `π_k` boundary clamps ride `hessian_diag`'s convention (`pi_jac=0` ⇒ the M-channel vanishes), so the channel is one-sided exactly where the assembled curvature is locally constant in `M_k` (landmine 3).

**Verification ladder.**
1. `sae_logdet_theta_adjoint_matches_dense_fd_ibp_map` (in-crate): Γ vs centered FD of `arrow_log_det` at the converged state, probing logits across distinct rows so the cross-row `M_k` coupling is exercised — single-logit FD captures both channels because `fixed_state_logdet` rebuilds `H`.
2. `sae_outer_rho_gradient_components_match_centered_fd_ibp_map` (`tests/sae_outer_gradient_fd_gate.rs`): the FULL outer-ρ gradient (explicit + traces + Occam + the #1006 correction) vs centered FD of the actual REML criterion (inner re-solved per ρ), with `lambda_sparse` as the active IBP prior weight. softmax and jumprelu arms added alongside.

**Landmine 2 (S̃(θ) pullback):** unchanged — the decoder-smoothness `∂g/∂λ_smooth` and `∂H/∂θ` already share the inner assembly's reparameterized roughness; no `S̃ const` assumption was introduced.
